### PR TITLE
Notifications: add nav bar button to mark all as read

### PR DESCRIPTION
**Pull request by ScoutHarris**
Ref: #17135 
WPKit PR: https://github.com/wordpress-mobile/WordPressKit-iOS/pull/476

In the Notifications view, a checkmark button is added to the nav bar to mark all unread notifications as read for the selected filter.

This is to merge Alp's trial project PRs to `trunk`.
- #17772
- #17789
- #17796
- #17808

To test:
- Go to Notifications.
- Verify a checkmark icon now appears in the nav bar to the left of the settings button.
- For filters that do not have unread notifications, the checkmark is disabled.
- For filters that do have unread notifications, the checkmark is enabled.
- Tip: if you don't have any unread notifications, swipe right on a notification row to mark it unread.

| read | unread |
|--------|-------|

- On a filter that has unread notifications, tap the checkmark button.
- Verify all the notifications in that filter are marked as read, and the button is disabled.
- Verify notifications in other filters are not.
- Note that, since the button applies to the selected filter, `All` and `Unread` will mark all unread notifications as read (thus affecting the other filters).

## Regression Notes
1. Potential unintended areas of impact
It's possible marking a single notification as read/unread could be affected.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Verified marking a single notification as read/unread still works as expected. And there are existing tests that passed.

3. What automated tests I added (or what prevented me from doing so)
Tests were added to WPKit.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Services/NotificationSyncMediator.swift [hunk 1]
```diff
@@ -257,15 +282,19 @@ class NotificationSyncMediator {
     /// Invalidates the local cache for the notification with the specified ID.
     ///
     func invalidateCacheForNotification(with noteID: String) {
+        invalidateCacheForNotifications(with: [noteID])
+    }
+
+    /// Invalidates the local cache for all the notifications with specified ID's in the array.
+    ///
+    func invalidateCacheForNotifications(with noteIDs: [String]) {
```
Comment (frosty) [inline WordPress/Classes/Services/NotificationSyncMediator.swift]: Nitpick but I'd probably switch the `with` for a `_` here, as I think it reads better when calling the method.

**Review by frosty — APPROVED** [review]
This works well for me, just left one nitpick.

UX-wise, I'm not sure if this is coming as a future change or anything, but I feel like it would be good for this feature to have either confirmation (like an action sheet alert controller asking you to confirm marking all as read), or an undo action available on the success notice. Particularly as it's not _super_ obvious what the checkmark will do before you tap it. Have either of these things been considered?

**Comment by alpavanoglu** [comment]
@frosty hi, I think that is a great point. I would say confirming is a better fit here than undo as it also provides an information what the button does before the action is taken. What do you think @ScoutHarris ?

**Comment by ScoutHarris** [comment]
Yea, I agree - we should have a confirmation. Thanks for noticing @frosty ! We'll get that added before merging.

**Comment by frosty** [comment]
I think given that this is feature flagged, it would be fine to merge this and add the alert in a follow-up PR if that's easier :)

**Comment by ScoutHarris** [comment]
> I think given that this is feature flagged, it would be fine to merge this and add the alert in a follow-up PR if that's easier

Weeeeellllll coincidentally there's some pod issues after the last release (🤞 to be resolved with [this PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/17810)), so I can't update this PR yet to resolve the file conflicts or it won't build. So we'll just let this marinate a while and see what Alp comes up with and when. 😄

**Pull request merged**